tough-ssm/src/error.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use snafu::{Backtrace, Snafu};
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The error type for this library.
+#[derive(Debug, Snafu)]
+#[snafu(visibility = "pub(crate)")]
+pub enum Error {
+    #[snafu(display("Unable to parse keypair: {}", source))]
+    KeyPairParse {
+        source: tough::error::Error,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Error creating AWS credentials provider: {}", source))]
+    RusotoCreds {
+        source: rusoto_credential::CredentialsError,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Unknown AWS region \"{}\": {}", region, source))]
+    RusotoRegion {
+        region: String,
+        source: rusoto_core::region::ParseRegionError,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Error creating AWS request dispatcher: {}", source))]
+    RusotoTls {
+        source: rusoto_core::request::TlsError,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "Failed to get aws-ssm://{}{}: {}",
+        profile.as_deref().unwrap_or(""),
+        parameter_name,
+        source,
+    ))]
+    SsmGetParameter {
+        profile: Option<String>,
+        parameter_name: String,
+        source: rusoto_core::RusotoError<rusoto_ssm::GetParameterError>,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Missing field in SSM response: {}", field))]
+    SsmMissingField {
+        field: &'static str,
```

Comment:
Same as above

Reply:
Sorry, same as what above?  If the profile doesn't include region, I think it'd be helpful to separately include region, and the parameter name.  Unless there's only one?  If it's obvious because there's only one, you can ignore me...

Reply:
Sorry if I wasn't clear, we won't have region here either.

In this context, we're telling the user that SSM didn't include something in it's response. [These are the response values](https://rusoto.github.io/rusoto/rusoto_ssm/struct.Parameter.html) I have to work with, which includes the parameter that we queried for (`Parameter.name`).  However, it's an `Option<String>` so it requires handling of its own, which gets weird in the case of `None`. It would be odd for the user to get an error that says "SSM returned a response to me that didn't include the name of the parameter we queried for". I think it's clearer to the user to get an error that says "SSM didn't return the value we queried for" (in this case `parameter` or `parameter.value`).

Reply:
Let me ask a different way - will we only have a single KeySource in a single region in any given run?  If so, then the user knows what parameter they requested, and we wouldn't need to include anything about it specifically in the error.  If not, they won't know which one failed, and it'd be helpful to include the parameter name and region in the error.

Separately, it was confusing to me to reuse the `SsmMissingField` error for two cases - we didn't get a parameter back at all, and we did get a parameter back but it was missing a value.  The first seems to indicate that the user specified the parameter incorrectly or are in the wrong region or something, where the second indicates that they specified it correctly but set the parameter incorrectly or there was an SSM service issue.

Reply:
I think the answer to your first question is "yes". The user supplies the SSM KeySource at the command line, so I think they'll know what parameter they requested.

As far as the `SsmMissingField` error is concerned, I feel it's clear enough without trying to diagnose the issue. "We got a response, but it didn't contain what we wanted".  If you feel strongly about it, I'm fine with adding additional error variants.

Reply:
> I think the answer to your first question is "yes". The user supplies the SSM KeySource at the command line, so I think they'll know what parameter they requested.

Is this only true with our current usage in tuftool, or does it apply to all usage of tough-ssm?  If we're moving this logic to tough, we have to think about use cases other than tuftool, and if that means there can be multiple `KeySource`s, then I think we have to consider clarity issues like this more carefully.  If that doesn't apply, we can resolve this.

Reply:
I've added the parameter name to the error to make it clearer.